File: 1125_SmallestSufficientTeam.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class SmallestSufficientTeam {
  public int[] smallestSufficientTeam(String[] req_skills, List<List<String>> people) {
    int m=people.size(), n=req_skills.length, target=0;
    Map<String, Integer> map = new HashMap<>();
    for (int i=0; i<n; ++i) {
      target += 1 << i;
      map.put(req_skills[i], i);
    }

    int[] pp = new int[m];
    for (int i=0; i<m; ++i) 
      for (String s : people.get(i))
        pp[i] += 1 << map.get(s);

    Map<Integer, List<Integer>> memo = new HashMap<>();
    memo.put(0, new ArrayList<>());
    while (true) {
      Map<Integer, List<Integer>> next = new HashMap<>();
      for (int k : memo.keySet()) {
        List<Integer> temp_list = memo.get(k);
        for (int i=0; i<m; ++i) 
          if (!temp_list.contains(i)) {
            int temp = k | pp[i];
            if (temp>k && !memo.containsKey(temp) && !next.containsKey(temp)) {
              List<Integer> res = new ArrayList<>(temp_list);
              res.add(i);
              if (temp==target) {
                int[] ans = new int[res.size()];
                Arrays.setAll(ans, res::get);
                return ans;
              }
              next.put(temp, res);
            }
          }
      }
      memo = next;
    }
  }
}

File: 1125_SmallestSufficientTeam.cpp
#include <local_leetcode.hpp>

class Solution {
  public:
    vector<int> smallestSufficientTeam(vector<string>& req_skills, vector<vector<string>>& people) {
      const int n = req_skills.size(); // 1 ~ 16
      const int m = people.size(); // 1 ~ 60

      unordered_map<string, uint16_t> skill_ids;
      for (int i = 0; i < n; ++i)
        skill_ids[req_skills[i]] = 1U << i;

      vector<uint16_t> ppl(m);
      for (int i = 0; i < m; ++i)
        for (const string& s : people[i])
          ppl[i] |= skill_ids[s];

      uint64_t dp[1 << 16]; // bits of skills, bits of people
      memset(dp, 0, sizeof(uint64_t) * (1 << n));

      queue<uint16_t> q;
      q.push(0);

      while (!q.empty()) {
        uint16_t u = q.front();
        q.pop();

        for (int i = 0; i < m; ++i) {
          uint16_t v = u | ppl[i];

          // target found
          if (v == static_cast<uint16_t>((1 << n) - 1)) {
            vector<int> res;

            uint64_t temp = dp[u];
            for (int j = 0; j < m; ++j)
              if ((temp >> j) & 1)
                res.push_back(j);

            res.push_back(i);
            
            return res;
          }

          // new skills found and the new combination
          if (v > u && dp[v] == 0) {
            dp[v] = dp[u] | (1ULL << i);
            q.push(v);
          }
        }
      }

      return {};
    }
};


int main() {
  EXECS(Solution::smallestSufficientTeam);
  return 0;
}